File: euclidean_gcd/C/euclidean_gcd.c
#include <stdio.h>

/*
Calculates GCD of two numbers a & b using the division-based Euclidean Algorithm
 */
int gcd_div(int a , int b) {
	while (b != 0) {
		int temp = b;
		b = a % b;
		a = temp;
	}
	return a;
}

/*
Calculates GCD of two numbers a & b using the recursive-based Euclidean Algorithm
 */
int gcd_rec(int a, int b) {
	if (a == 0)
		return b;
	return gcd_rec(b % a, a);
}

/*
Calculates GCD of two numbers a & b using the Recursive-based Extended Euclidean Algorithm
 */
int gcd_extended(int a, int b, int *x, int *y) {
	if (a == 0) {
		*x = 0;
		*y = 1;
		return b;
	}
	int x1, y1;
	int result = gcd_extended(b % a, a, &x1, &y1);
	*x = y1 - (b / a) * x1;
	*y = x1;
	return result;
}

// Main Method
int main() {
	int a = 20, b = 30, x, y;
	printf("Division: GCD(%d, %d) = %d\n", a, b, gcd_div(a, b));

	printf("Recursive: GCD(%d, %d) = %d\n", a, b, gcd_rec(a, b));

	int gcd_ext = gcd_extended(a, b, &x, &y);
	printf("Extended: GCD(%d, %d) = %d\n", a, b, gcd_ext);
}

File: euclidean_gcd/Java/euclidean_gcd.java
import java.util.*;


public class euclidean_gcd {

    /*
    * Calculates GCD of two numbers a & b using the division-based
    * Euclidean Algorithm
    */
    public static long euclideanGcdDivision(long x, long y) {
        while(y != 0) {
            long temp =y;
            y=x%y;
            x=temp;
        }
        return x;
    }

    /*
    * Calculates GCD of two numbers a & b using the Recursive-based
    * Euclidean Algorithm
    */
    public static long euclideanGcdRecursive(long x, long y) {
        if (x == 0)
            return y;
        return euclideanGcdRecursive(y%x, x);
    }

    /*
    * Calculates GCD of two numbers a & b using the Recursive-based Extended
    * Euclidean Algorithm
    */
    public void extendedGcd(long a, long b) {
        long gcd_rec = euclideanGcdRecursive(a,b);
        System.out.println("Extended_Recursive: GCD of " +a+ " & " +b+ " is : " +gcd_rec);
        long x = 0, y = 1, Prev_x = 1, Prev_y = 0, temp;
        while (b != 0) {
            long q = a / b;
            long r = a % b;
            a = b;
            b = r;
            temp = x;
            x = Prev_x - q * x;
            Prev_x = temp;
            temp = y;
            y = Prev_y - q * y;
            Prev_y = temp;
        }
        System.out.println("Roots  x : "+ Prev_x +" y :"+ Prev_y);
    }

    public static void main (String[] args) {
        long a = 20;
        long b = 30;

        long gcd_div = euclideanGcdDivision(a, b);
        System.out.println("Division: GCD of " +a+ " & " +b+ " is : " + gcd_div);

        long gcd_rec = euclideanGcdRecursive(a, b);
        System.out.println("Recursive: GCD of " +a+ " & " +b+ " is : " + gcd_rec);

        euclidean_gcd obj = new euclidean_gcd();
        obj.extendedGcd(a, b);
    }
}
